text/utilities/0000-health-checker-buildpack.md

```diff
@@ -0,0 +1,70 @@
+
+Having a buildpack that installs a health checker allows for HTTP-based health checks when running on Docker. In addition, it could allow for other types of health checkers for example if your application does not use HTTP.
+
+## Detailed Explanation
+
+1. A health checker buildpack has been developed. It will be contributed to Paketo.
+   1. By default, no health checker will be installed by the buildpack.
+   2. If a user sets `BP_HEALTH_CHECKER_ENABLED=true` then the buildpack will detect and contribute a health check binary to the `$PATH`.
+   3. A process type of `health-check` will be contributed with the command to run the health checker. This may be invoked by the user rather than needing the full command.
+
+## Rationale and Alternatives
+
+1. A user can create a custom stack and include a tool like `curl` or `wget`. This is a lot of additional work, plus `curl` and `wget` are overkill for a simple HTTP health check. To reduce footprint, they are not recommended.
+
+## Implementation
+
+The new Health Checker buildpack will participate if all the following conditions are met
+
+* A user sets `BP_HEALTH_CHECKER_ENABLED=true`
+
+When conditions are met, the buildpack will do the following:
+
+* Contributes a health checker to a layer marked `launch` with command on `$PATH`
+* Contributes a process type called `health-check` which will execute the installed health checker.
+
+The Health Checker buildpack will optionally be added to the language family buildpacks that wish to include support for it. The primary targets will be the Java and Rust buildpacks.
+
+### Example
+
+A user can then utilize the health checker like this:
+
+1. For a Spring Boot app, run `pack build apps/maven -p target/demo-0.0.1-SNAPSHOT.jar -b paketo-buildpacks/java -b docker.io/paketo-buildpacks/health-checker -e BP_HEALTH_CHECKER_ENABLED=true -e BP_HEALTH_CHECKER_THC_ARGS='8080 /actuator/health'` (it won't be necessary to include the `-b` arguments once the buildpack has been integrated with language family buildpacks). You'll see output like this at the end. It's installing the health checker and setting a process type.
+
+    ```
+    Paketo Health Checker Buildpack vDEVELOPMENT
+      https://github.com/paketo-buildpacks/health-checker
+      Build Configuration:
+        $BP_HEALTH_CHECKER_DEPENDENCY  thc                    which health checker to contribute
+        $BP_HEALTH_CHECKER_ENABLED     true                   contributes a health checker if enabled
+        $BP_HEALTH_CHECKER_THC_ARGS    8080 /actuator/health  arguments passed to tiny-health-checker, if used
```

Comment:
I would like to see a discussion of the implementation. I.E what does the buildpack do with `BP_HEALTH_CHECKER_THC_ARGS`? Are they added as `args` in the process definition? That would make them hard to override at runtime without the changes proposed in currently in-flight [process refactor RFC](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md) but might work very nicely once it lands.

The other option would be to configure the `thc` via runtime environment variables (with defaults set the buildpack).

Reply:
This is a fair point. 

The current stance is that we use `BP_HEALTH_CHECKER_THC_ARGS` (`BP_*` specifically) because the arguments get set at build time and are not intended to be changed at runtime. They are added directly to the process type arguments list.

The thought process is that the application developer will know what health check they have exposed and what is a meaningful health check, thus they should define it and make it part of the image. That makes things easier for the operator at runtime. They don't have to think about what health check to use, they just run `/cnb/process/health-check` and it should work.

I'm not necessarily opposed to adding runtime configuration, if others think it will be of value. I'm not sure that using `BPL_*` makes sense though. There would be some complications to make `BPL_*` work. The health checker isn't going to natively understand them, so you'd need an exec.d or you'd need to tightly couple the health checker to the buildpack (ie. the health checker would need to look at `BPL_...`). I'm also to sure how you'd get around the limitations exec.d where it cannot modify the process types. Essentially, you'd need a health checker that is fully configurable via env variables. We could do that to `thc`, but it would make using other off-the-shelf health checkers harder. In addition, if we require the health checker to be configurable via env variables then we could probably just not use `BPL_*` to begin with and use the direct health checker config.

Besides that, there is a backdoor built-in here that would allow the user to override the health check and that is to just invoke the health checker directly, instead of through the `/cnb/process/health-check` launcher. The main benefit of using the launcher is so the operator doesn't need to know how to run the health checker, but if the operator wants to override the settings at runtime then it seems reasonable to expect them to look up the arguments on how to run the health checker. The only inconvenience would be that they'd need the full path to the health checker or to invoke the launcher so that `$PATH` is set up correctly.

Reply:
>The other option would be to configure the thc via runtime environment variables (with defaults set the buildpack).

What would be the path to make this work in a similar way to the use of args. Right now, we can bake the args into the image, so it's just a single command to run, `/cnb/process/health-check`. 

You can't set env variables on a process type, so while we could add a process type, it would have less utility. The user would need to configure the env variables still. 

The only way I can think to bake in the env variables so that there is a default "it just works" scenario is by pairing this with the environment-variables buildpack. Is that what you were thinking too? or am I overlooking a better way to do that?

Reply:
Excellent. Let me make some adjustments with this in mind.